efi/secureboot_namespace_rules.go

```diff
@@ -19,8 +19,120 @@
+func withAuthority(subject, subjectKeyId []byte, publicKeyAlgorithm x509.PublicKeyAlgorithm) secureBootNamespaceOption {
+	return func(ns *secureBootNamespaceRules) {
+		ns.authorities = append(ns.authorities, &secureBootAuthorityIdentity{
+			subject:            subject,
+			subjectKeyId:       subjectKeyId,
+			publicKeyAlgorithm: publicKeyAlgorithm})
+	}
+}
+
+// withImageRule adds the specified rule to a secureBootNamespaceRules.
+func withImageRule(name string, match imagePredicate, create newImageLoadHandlerFn) secureBootNamespaceOption {
+	return func(ns *secureBootNamespaceRules) {
+		ns.rules = append(ns.rules, newImageRule(name, match, create))
+	}
+}
+
+type secureBootNamespaceOption func(*secureBootNamespaceRules)
+
+// secureBootNamespaceRules is used to construct an imageLoadHandler from a
+// peImageHandle using a set of rules that are scoped to a secure boot
+// hierarchy.
+type secureBootNamespaceRules struct {
+	authorities []*secureBootAuthorityIdentity
+	*imageRules
+}
+
+// newSecureBootNamespaceRules constructs a secure boot namespace with the specified
+// options.
+func newSecureBootNamespaceRules(name string, options ...secureBootNamespaceOption) *secureBootNamespaceRules {
+	out := &secureBootNamespaceRules{
+		imageRules: newImageRules(name + " secure boot namespace"),
+	}
+	for _, option := range options {
+		option(out)
+	}
+	return out
+}
+
+func (r *secureBootNamespaceRules) AddAuthorities(certs ...*x509.Certificate) {
+	for _, cert := range certs {
+		found := false
+		for _, authority := range r.authorities {
+			if bytes.Equal(authority.subject, cert.RawSubject) &&
+				bytes.Equal(authority.subjectKeyId, cert.SubjectKeyId) &&
+				authority.publicKeyAlgorithm == cert.PublicKeyAlgorithm {
+				found = true
+				break
+			}
+		}
+		if !found {
+			r.authorities = append(r.authorities, &secureBootAuthorityIdentity{
+				subject:            cert.RawSubject,
+				subjectKeyId:       cert.SubjectKeyId,
+				publicKeyAlgorithm: cert.PublicKeyAlgorithm,
+			})
+		}
+	}
+}
+
+func (r *secureBootNamespaceRules) NewImageLoadHandler(image peImageHandle) (imageLoadHandler, error) {
+	// This may return no signatures, but that's ok - in the case, we just return
+	// errNoHandler.
+	sigs, err := image.SecureBootSignatures()
+	if err != nil {
+		// Reject any image with a badly formed security directory entry
+		return nil, xerrors.Errorf("cannot obtain secure boot signatures: %w", err)
+	}
+
+	for _, authority := range r.authorities {
+		cert := &x509.Certificate{
+			RawSubject:         authority.subject,
+			SubjectKeyId:       authority.subjectKeyId,
+			PublicKeyAlgorithm: authority.publicKeyAlgorithm}
+		for _, sig := range sigs {
+			if !sig.CertLikelyTrustAnchor(cert) {
+				continue
+			}
+
+			handler, err := r.imageRules.NewImageLoadHandler(image)
+			if err != nil {
+				return nil, err
+			}
+
+			if v, ok := handler.(vendorAuthorityGetter); ok {
+				certs, err := v.VendorAuthorities()
```

Comment:
are authorities coming from a handler from a rule a subset of the authorities in r.authorities? wouldn't it make more sense the always add rules with some authorities together instead of separately? I'm looking at the separate with* helpers above

Reply:
I'm not sure I understand the question. The comment for vendorAuthorityGetter might help a bit, and it might be useful to look at how this is put together in the next PR as well:

https://github.com/snapcore/secboot/blob/e4c2c15fb92e8673e28d68343fcc59d868e08172/efi/image_rules_defs.go